specs/core/1_custody-game.md

```diff
@@ -288,32 +329,93 @@ Add the following operations to the per-block processing, in order the given bel
+Replace process_custody_reveal from phase 0 with this function:
 
 ```python
 def process_custody_reveal(state: BeaconState,
-                           reveal: CustodyKeyReveal) -> None:
+                           reveal: RandaoKeyReveal) -> None:
+
+    revealed_validator = state.validator_registry[reveal.revealed_index]
+
+    assert reveal.epoch == get_randao_epoch_for_custody_period( revealed_validator.next_custody_reveal_period, reveal.revealed_index)
+    assert revealed_validator.next_custody_reveal_period < get_validators_custody_reveal_period(state, reveal.revealed_index)
+
+    # Revealed validator is active or exited, but not withdrawn
+    assert is_slashable_validator(revealer, get_current_epoch(state))
+
+    # Decrement max reveal lateness if response is timely
+    if revealed_validator.next_custody_reveal_period == get_validators_custody_reveal_period(state, reveal.revealer_index) - 2:
+            revealer.max_reveal_lateness -=  MAX_REVEAL_LATENESS_DECREMENT
+    revealed_validator.max_reveal_lateness = max(revealed_validator.max_reveal_lateness, get_validators_custody_reveal_period(state, reveal.revealed_index) - revealed_validator.next_custody_reveal_period)
+
+    # Process reveal
+    revealed_validator.next_custody_reveal_period += 1
+
+    # Reward Block Preposer
+    proposer_index = get_beacon_proposer_index(state)
+    increase_balance(state, proposer_index, base_reward(state, index) // MINOR_REWARD_QUOTIENT)
+```
+
+##### Randao key reveals
+
+Verify that `len(block.body.randao_key_reveals) <= MAX_RANDAO_KEY_REVEALS`.
+
+For each `reveal` in `block.body.randao_key_reveal`, run the following function:
+
+```python
+def process_randao_key_reveal(state: BeaconState,
+                              reveal: RandaoKeyReveal) -> None:
+    """
+    Process ``RandaoKeyReveal`` operation.
+    Note that this function mutates ``state``.
+    """
+    # Verify signature correctness
     assert verify_custody_key(state, reveal)
-    revealer = state.validator_registry[reveal.revealer_index]
-    current_custody_period = epoch_to_custody_period(get_current_epoch(state))
 
-    # Case 1: non-masked non-punitive non-early reveal
     if reveal.mask == ZERO_HASH:
-        assert reveal.period == epoch_to_custody_period(revealer.activation_epoch) + revealer.custody_reveal_index
-        # Revealer is active or exited
-        assert is_active_validator(revealer, get_current_epoch(state)) or revealer.exit_epoch > get_current_epoch(state)
-        revealer.custody_reveal_index += 1
-        revealer.max_reveal_lateness = max(revealer.max_reveal_lateness, current_custody_period - reveal.period)
-        proposer_index = get_beacon_proposer_index(state)
-        increase_balance(state, proposer_index, base_reward(state, index) // MINOR_REWARD_QUOTIENT)
-
-    # Case 2: masked punitive early reveal
+        # Handle the case of a regular custody reveal
+        process_custody_reveal(state, reveal)
```

Comment:
Maybe for symmetry put the logic below this into a separate function  `process_randao_key_punitive_reveal` and have the top-level function just be:

```
    revealed_validator = state.validator_registry[reveal.revealed_index]

    if reveal.mask == ZERO_HASH: 
        pubkeys = [revealed_validator.pubkey]
        message_hashes = [hash_tree_root(reveal.epoch)]
    else:
        masker = state.validator_registry[reveal.masker_index]
        pubkeys = [revealed_validator.pubkey, masker.pubkey]
        message_hashes = [
            hash_tree_root(reveal.epoch),
            reveal.mask,
        ]

    assert bls_verify_multiple(
        pubkeys=pubkeys,
        message_hashes=message_hashes,
        signature=reveal.reveal,
        domain=get_domain(
            state=state,
            domain_type=DOMAIN_RANDAO,
            message_epoch=reveal.epoch,
        ),
    )
    if reveal.mark == ZERO_HASH:
        process_custody_reveal(state, reveal)
    else:
        process_randao_key_punitive_reveal(state, reveal)
```

This, plus the suggestion below, would allow the `verify_custody_key` function to be cut out entirely.

Reply:
Though honestly if we have this parallel structure the right approach may be to just have two objects, a `CustodyKeyReveal` and a `EarlyReveal`, where only the latter has the masking mechanism and the former does not (the former would just be a thin wrapper `{validator_index: uint64, custody_key: bytes96}`). This way we don't need the empty mask fields in legitimate key reveals, and the code can be separated out more instead of being an if statement that runs totally different code depending on what role the object is serving.

Reply:
I did this using two objects now. Pls review and check if this is your preferred method :)